### ISSUE TITLE
Tests fail with Python 3.8

### ISSUE THREAD

**Issue by blais** [beancount/beancount#410]
**[Original report](https://bitbucket.org/blais/beancount/issue/410) by Dr. Tobias Quathamer (GitHub: [toddy15](https://github.com/toddy15)).**

----------------------------------------

Hi,

I’m helping to maintain the Debian package of beancount, and the tests have recently begun to fail, after we switched from Python 3.7 to Python 3.8. You can see the full build log here:

[https://buildd.debian.org/status/fetch.php?pkg=beancount&arch=amd64&ver=2.2.3-1%2Bb1&stamp=1574207855&raw=0](https://buildd.debian.org/status/fetch.php?pkg=beancount&arch=amd64&ver=2.2.3-1%2Bb1&stamp=1574207855&raw=0)

The first failing test is this:

`beancount/ingest/identify_test.py::TestScriptIdentify::test_identify FAILED [ 20%]`

Afterwards, the tests break with a segmentation fault error:

```
beancount/parser/lexer_test.py::TestLexerErrors::test_lexer_invalid_token PASSED [ 55%]
beancount/parser/lexer_test.py::TestLexerUnicode::test_bytes_encoded_latin1 PASSED [ 55%]
beancount/parser/lexer_test.py::TestLexerUnicode::test_bytes_encoded_latin1_invalid PASSED [ 55%]
beancount/parser/lexer_test.py::TestLexerUnicode::test_bytes_encoded_utf16 Segmentation fault
E: pybuild pybuild:341: test: plugin distutils failed with: exit code=139: cd /<<PKGBUILDDIR>>/.pybuild/cpython3_3.8/build; python3.8 -m pytest -v
dh_auto_test: pybuild --test --test-pytest -i python{version} -p "3.8 3.7" returned exit code 13
make: *** [debian/rules:13: binary-arch] Error 255
dpkg-buildpackage: error: debian/rules binary-arch subprocess returned exit status 2
```

‌

The corresponding bug in Debian is [https://bugs.debian.org/943608](https://bugs.debian.org/943608), however, there’s not much information in the bug log \(yet\). Could you please take a look at this? 

‌

Regards,

Tobias

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

I upgraded to 3.8 this morning on my Ubuntu 18.04.3 LTS box and I’m not reproducing that specific error.

However, it’s bringing up a segfault in the lexer.

I have to investigate.

**Comment by blais** [comment]
**Original comment by Daniele Nicolodi (Bitbucket: [daniele](https://bitbucket.org/daniele), GitHub: [dnicolodi](https://github.com/dnicolodi)).**

----------------------------------------

I tried to investigate the segfault and I came to the conclusion that it is a a CPython 3.8 bug. I reported it here: [https://bugs.python.org/issue38913](https://bugs.python.org/issue38913)

**Comment by blais** [comment]
**Original comment by Daniele Nicolodi (Bitbucket: [daniele](https://bitbucket.org/daniele), GitHub: [dnicolodi](https://github.com/dnicolodi)).**

----------------------------------------

Pull request #139 has a \(fairly clean and straigthforward\) workaround for the Python 3.8 issue causing the segmentation fault. I cannot reproduce the failing test. Having the complete test report will probably help understand what is going wrong there.

**Comment by blais** [comment]
**Original comment by Martin Blais (Bitbucket: [blais](https://bitbucket.org/blais), GitHub: [blais](https://github.com/blais)).**

----------------------------------------

parser: Work around python3.8 bug causing a segmentation fault

See https://bugs.python.org/issue38913

Fixes #410
